File: web/src/api/fund.ts
/** 基金相关 API */
import client from './client'

export interface FundHistoryItem {
  date: string
  nav: number | null
  pct_change: number | null
}

export interface FundItem {
  code: string
  name: string
  fund_type: string
  company: string
  manager: string
  remark: string
  added_at: string
  nav: number | null
  nav_date: string
  est_nav: number | null
  est_pct: number | null
  update_time: string
  history: FundHistoryItem[]
  nav_change_pct: number | null
}

export interface FundSearchItem {
  code: string
  name: string
  fund_type: string
  company: string
}

/** 获取自选基金列表 */
export async function getFundWatchlist(): Promise<FundItem[]> {
  const { data } = await client.get('/fund/watchlist')
  return data.data
}

/** 添加自选基金 */
export async function addFundWatchlist(code: string, remark?: string): Promise<any> {
  return client.post(`/fund/watchlist/${code}`, null, { params: { remark } })
}

/** 移除自选基金 */
export async function removeFundWatchlist(code: string): Promise<any> {
  return client.delete(`/fund/watchlist/${code}`)
}

/** 搜索基金 */
export async function searchFund(keyword: string): Promise<FundSearchItem[]> {
  const { data } = await client.get('/fund/search', { params: { q: keyword } })
  return data
}

/** 基金详情 */
export interface FundDetail {
  code: string
  name: string
  full_name: string
  fund_type: string
  company: string
  manager: string
  setup_date: string
  scale: string
  benchmark: string
  strategy: string
  nav: number | null
  nav_date: string
  est_nav: number | null
  est_pct: number | null
  update_time: string
}

/** 基金历史净值数据点 */
export interface FundNavPoint {
  date: string
  nav: number
  pct_change: number | null
}

/** 获取基金详情 */
export async function getFundDetail(code: string): Promise<FundDetail> {
  const { data } = await client.get(`/fund/detail/${code}`)
  return data
}

/** 获取基金历史净值 */
export async function getFundNavHistory(
  code: string, period: string = '1年'
): Promise<{ code: string; period: string; data: FundNavPoint[] }> {
  const { data } = await client.get(`/fund/nav-history/${code}`, { params: { period } })
  return data
}

// ---------- 基金回测 ----------

export interface FundStrategyInfo {
  id: string
  name: string
  description: string
  params_desc: string
}

export interface StrategyParams {
  interval_days?: number
  amount?: number
  drop_pct?: number
  levels?: { drop_pct: number; amount: number }[]
  target_value?: number
  rebalance_days?: number
  grid_pct?: number
  amount_per_grid?: number
  take_profit_pct?: number
}

export interface FundBacktestRequest {
  code: string
  start_date: string
  end_date?: string
  initial_capital: number
  strategy: string
  params: StrategyParams
}

export interface FundTradeRecord {
  date: string
  type: string
  nav: number
  shares: number
  amount: number
  fee: number
  cash_after: number
  position_value_after: number
  reason: string
}

export interface FundEquityPoint {
  date: string
  nav: number
  total: number
  cash: number
  position_value: number
  shares: number
  cost_basis: number
}

export interface FundBacktestStats {
  initial_capital: number
  total_invested: number
  final_value: number
  total_return: number
  total_return_pct: number
  annualized_return_pct: number
  max_drawdown_pct: number
  max_drawdown_date: string
  sharpe_ratio: number | null
  num_trades: number
  num_buys: number
  num_sells: number
  avg_buy_amount: number
  final_shares: number
  final_nav: number
}

export interface FundBacktestResponse {
  code: string
  name: string
  strategy: string
  params: StrategyParams
  stats: FundBacktestStats
  trades: FundTradeRecord[]
  equity_curve: FundEquityPoint[]
}

/** 获取回测策略列表 */
export async function getFundBacktestStrategies(): Promise<FundStrategyInfo[]> {
  const { data } = await client.get('/fund/backtest/strategies')
  return data
}

/** 运行基金回测 */
export async function runFundBacktest(req: FundBacktestRequest): Promise<FundBacktestResponse> {
  const { data } = await client.post('/fund/backtest/run', req)
  return data
}

File: server/routers/fund.py
"""基金路由"""
from fastapi import APIRouter, Query
from pydantic import BaseModel
from typing import List, Optional

from server.services import fund_service
from server.services import fund_backtest_service
from server.models.fund_backtest import FundBacktestRequest, FundBacktestResponse

router = APIRouter(prefix="/fund", tags=["基金"])


class FundHistoryItem(BaseModel):
    date: Optional[str] = ''
    nav: Optional[float] = None
    pct_change: Optional[float] = None


class FundItem(BaseModel):
    code: str
    name: str
    fund_type: Optional[str] = ''
    company: Optional[str] = ''
    manager: Optional[str] = ''
    remark: Optional[str] = ''
    tags: List[str] = []
    added_at: Optional[str] = ''
    nav: Optional[float] = None
    nav_date: Optional[str] = ''
    est_nav: Optional[float] = None
    est_pct: Optional[float] = None
    update_time: Optional[str] = ''
    history: List[FundHistoryItem] = []
    nav_change_pct: Optional[float] = None


class FundWatchlistResponse(BaseModel):
    data: List[FundItem]


class FundSearchItem(BaseModel):
    code: str
    name: str
    fund_type: str = ''
    company: str = ''


class FundDetailResponse(BaseModel):
    code: str
    name: str
    full_name: Optional[str] = ''
    fund_type: Optional[str] = ''
    company: Optional[str] = ''
    manager: Optional[str] = ''
    setup_date: Optional[str] = ''
    scale: Optional[str] = ''
    benchmark: Optional[str] = ''
    strategy: Optional[str] = ''
    nav: Optional[float] = None
    nav_date: Optional[str] = ''
    est_nav: Optional[float] = None
    est_pct: Optional[float] = None
    update_time: Optional[str] = ''


class FundNavPoint(BaseModel):
    date: str
    nav: float
    pct_change: Optional[float] = None


class FundNavHistoryResponse(BaseModel):
    code: str
    period: str
    data: List[FundNavPoint]


@router.get("/watchlist", response_model=FundWatchlistResponse)
async def get_watchlist(cache_minutes: int = Query(2, description="估值缓存分钟数")):
    """获取自选基金列表（含估值，优先使用缓存）"""
    data = fund_service.get_watchlist(cache_minutes)
    return {"data": data}


@router.post("/watchlist/refresh")
async def refresh_estimations():
    """强制刷新所有自选基金估值"""
    return fund_service.refresh_estimations()


@router.post("/watchlist/{code}")
async def add_watchlist(code: str, remark: str = Query("", description="备注")):
    """添加自选基金"""
    result = fund_service.add_watchlist(code, remark)
    return result


@router.delete("/watchlist/{code}")
async def remove_watchlist(code: str):
    """移除自选基金"""
    result = fund_service.remove_watchlist(code)
    return result


@router.put("/watchlist/{code}/tags")
async def update_tags(code: str, tags: str = Query(..., description="标签，逗号分隔")):
    """更新基金标签"""
    return fund_service.update_tags(code, tags)


@router.get("/search", response_model=List[FundSearchItem])
async def search_fund(q: str = Query(..., min_length=1, description="搜索关键词")):
    """搜索基金（代码或名称）"""
    return fund_service.search_fund(q)


@router.get("/detail/{code}", response_model=FundDetailResponse)
async def get_fund_detail(code: str):
    """获取基金详情（基本信息 + 最新估值）"""
    detail = fund_service.get_fund_detail(code)
    if not detail:
        from fastapi import HTTPException
        raise HTTPException(status_code=404, detail=f"基金 {code} 不存在")
    return detail


@router.get("/nav-history/{code}", response_model=FundNavHistoryResponse)
async def get_fund_nav_history(
    code: str,
    period: str = Query('1年', description="时间周期: 1月/3月/6月/1年/3年/5年/今年来/成立来"),
):
    """获取基金历史净值数据"""
    data = fund_service.get_fund_nav_history(code, period)
    return {'code': code, 'period': period, 'data': data}


@router.get("/backtest/strategies")
async def get_backtest_strategies():
    """获取基金回测策略列表"""
    return fund_backtest_service.STRATEGIES


@router.post("/backtest/run", response_model=FundBacktestResponse)
async def run_fund_backtest(req: FundBacktestRequest):
    """运行基金回测"""
    try:
        return fund_backtest_service.run_fund_backtest(req)
    except ValueError as e:
        from fastapi import HTTPException
        raise HTTPException(status_code=400, detail=str(e))
